core/server/worker/src/main/java/alluxio/worker/block/UfsInputStreamManager.java

```diff
@@ -0,0 +1,343 @@
+      } else {
+        resources = new UfsInputStreamIds();
+        mFileIdToInputStreamIds.put(fileId, resources);
+      }
+    }
+
+    synchronized (resources) {
+      long nextId = UNAVAILABLE_RESOURCE_ID;
+      CachedSeekableInputStream inputStream = null;
+      if (reuse) {
+        // find the next available input stream from the cache
+        for (long id : resources.availableIds()) {
+          inputStream = mUnderFileInputStreamCache.getIfPresent(id);
+          if (inputStream != null) {
+            nextId = id;
+            LOG.debug("Reused the under file input stream resource of {}", nextId);
+            // for the cached ufs instream, seek to the requested position
+            inputStream.seek(openOptions.getOffset());
+            break;
+          }
+        }
+      }
+      // no cached input stream is available, open a new one
+      if (nextId == UNAVAILABLE_RESOURCE_ID) {
+        nextId = IdUtils.getRandomNonNegativeLong();
+        final long newId = nextId;
+        try {
+          inputStream = mUnderFileInputStreamCache.get(nextId, () -> {
+            SeekableUnderFileInputStream ufsStream = (SeekableUnderFileInputStream) ufs.open(path,
+                OpenOptions.defaults().setOffset(openOptions.getOffset()));
+            LOG.debug("Created the under file input stream resource of {}", newId);
+            return new CachedSeekableInputStream(ufsStream, newId, fileId, path);
+          });
+        } catch (ExecutionException e) {
+          LOG.warn("Failed to create a new cached ufs instream of file id {} and path {}", fileId,
+              path);
+          // fall back to a ufs creation.
+          return ufs.open(path, OpenOptions.defaults().setOffset(openOptions.getOffset()));
+        }
+      }
+
+      // mark the input stream id as acquired
+      resources.acquire(nextId);
+      return inputStream;
+    }
+  }
+
+  private boolean isCachingEnabled() {
+    return Configuration.getBoolean(PropertyKey.WORKER_UFS_INSTREAM_CACHE_ENABLED);
+  }
+
+  /**
+   * The metadata of the input streams associated with an under storage file that tracks which input
+   * streams are in-use or available. Each input stream is identified by a unique id.
+   */
+  @ThreadSafe
+  private static class UfsInputStreamIds {
+    private final Set<Long> mInUse;
+    private final Set<Long> mAvailable;
+
+    /**
+     * Creates a new {@link UfsInputStreamIds}.
+     */
+    UfsInputStreamIds() {
+      mInUse = new HashSet<>();
+      mAvailable = new HashSet<>();
+    }
+
+    /**
+     * @return a view of the available input stream ids
+     */
+    Set<Long> availableIds() {
+      return Collections.unmodifiableSet(mAvailable);
+    }
+
+    /**
+     * Marks an input stream as acquired.
+     *
+     * @param id the id of the input stream
+     */
+    synchronized void acquire(long id) {
+      Preconditions.checkArgument(!mInUse.contains(id), id + " is already in use");
```

Comment:
```java
Preconditions.checkArgument(!mInUse.contains(id), "%s is already in use", id);
```
to avoid unnecessary string calculation on expected cases